gax/src/test/java/com/google/api/gax/rpc/mtls/AbstractMtlsTransportChannelTest.java

```diff
@@ -0,0 +1,93 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.api.gax.rpc.mtls;
+
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+import com.google.api.gax.rpc.mtls.MtlsProvider.MtlsEndpointUsagePolicy;
+import com.google.api.gax.rpc.testing.FakeMtlsProvider;
+import java.io.IOException;
+import org.junit.Test;
+
+public abstract class AbstractMtlsTransportChannelTest {
+  /**
+   * Returns the mTLS object from the created transport channel. mTLS object is created with mTLS
+   * keystore. For HttpJsonTransportChannel, the mTLS object is the SslContext; for
+   * GrpcTransportChannel, the mTLS object is the ChannelCredentials. The transport channel is mTLS
+   * if and only if the related mTLS object is not null.
+   */
+  protected abstract Object getMtlsObjectFromTransportChannel(MtlsProvider provider)
+      throws IOException;
+
+  @Test
+  public void testNotUseClientCertificate() throws IOException {
+    MtlsProvider provider =
+        new FakeMtlsProvider(false, MtlsEndpointUsagePolicy.AUTO, null, "", false);
+    assertNull(getMtlsObjectFromTransportChannel(provider));
+  }
+
+  @Test
+  public void testUseClientCertificate() throws IOException {
+    MtlsProvider provider =
+        new FakeMtlsProvider(
+            true,
+            MtlsEndpointUsagePolicy.AUTO,
+            FakeMtlsProvider.createTestMtlsKeyStore(),
+            "",
+            false);
+    assertNotNull(getMtlsObjectFromTransportChannel(provider));
+  }
+
+  @Test
+  public void testNoClientCertificate() throws IOException {
+    MtlsProvider provider =
+        new FakeMtlsProvider(true, MtlsEndpointUsagePolicy.AUTO, null, "", false);
+    assertNull(getMtlsObjectFromTransportChannel(provider));
+  }
+
+  @Test
+  public void testGetKeyStoreThrows() {
+    // Test the case where provider.getKeyStore() throws.
+    MtlsProvider provider =
+        new FakeMtlsProvider(true, MtlsEndpointUsagePolicy.AUTO, null, "", true);
+    try {
+      getMtlsObjectFromTransportChannel(provider);
+      fail("should throw and exception");
+    } catch (IOException e) {
+      assertTrue(
+          "expected to fail with exception",
```

Comment:
the optional message argument should explain what went wrong if we got here and the assert failed, not what happened if this line isn't reached

Reply:
Done.